tfjs-backend-webgpu/src/multinomial_webgpu.ts

```diff
@@ -0,0 +1,77 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =============================================================================
+ */
+
+import {getMainHeaderString as main, WebGPUProgram} from './webgpu_program';
+import {computeDispatch, flatDispatchLayout} from './webgpu_util';
+
+export class MultinomialProgram implements WebGPUProgram {
+  variableNames: string[] = ['probs'];
+  outputShape: number[] = [];
+  shaderKey: string;
+  dispatchLayout: {x: number[]};
+  dispatch: [number, number, number];
+  uniforms = 'seed : f32, numOutcomes: i32,';
+  workgroupSize: [number, number, number] = [64, 1, 1];
+  size = true;
+
+  constructor(batchSize: number, numSamples: number) {
+    this.outputShape = [batchSize, numSamples];
+    this.dispatchLayout = flatDispatchLayout(this.outputShape);
+    this.dispatch = computeDispatch(
+        this.dispatchLayout, this.outputShape, this.workgroupSize);
+
+    this.shaderKey = 'multinomial';
+  }
+
+  getUserCode(): string {
+    const userCode = `
+    //Based on the work of Dave Hoskins
+    //https://www.shadertoy.com/view/4djSRW
+    fn random (seed : f32, resultUV : vec2<f32>) -> f32 {
+      let HASHSCALE1 = 443.8975;
+      let p = resultUV * seed;
+      var p3  = fract(vec3<f32>(p.xyx) * HASHSCALE1);
+      p3 = p3 + dot(p3, p3.yzx + 19.19);
+      return fract((p3.x + p3.y) * p3.z);
+    }
+
+    ${main('index')} {
+      if (index < uniforms.size) {
+        let coords = getOutputCoords();
+        let batch = coords[0];
+
+        let resUV = vec2<f32>(f32(coords[1]) / f32(uniforms.outShape[1]),
+            f32(coords[0]) / f32(uniforms.outShape[0]));
+        let r = random(uniforms.seed, resUV);
+        var cdf = 0.0;
+        for (var i = 0; i < uniforms.numOutcomes; i = i + 1) {
```

Comment:
In WebGL impl, it's (uniforms.numOutcomes -1) here.

Reply:
Thanks for finding the error.

Reply:
Is it easy to add a case to catch such kind of error?

Reply:
`r` is random value, I think it is difficult to design the case.